ToolBox/PsychSignalDataConverter/PsychSignalDataDownloader.cs

```diff
@@ -0,0 +1,149 @@
+        private readonly string _dataSource;
+        
+        /// <summary>
+        /// Base URL for the psychsignal API
+        /// </summary>
+        private readonly string _baseUrl = "https://api.psychsignal.com/v2";
+
+        /// <summary>
+        /// Destination we will write raw data to
+        /// </summary>
+        private readonly string _rawDataDestination;
+        
+        /// <summary>
+        /// Maximum amount of retries per data hour 
+        /// </summary>
+        public int MaxRetries = 5;
+        
+        /// <summary>
+        /// Downlods data from psychsignal
+        /// </summary>
+        /// <param name="rawDataDestination">Directory we write raw data to</param>
+        /// <param name="apiKey">PsychSignal API key</param>
+        /// <param name="dataSource">Data source (e.g. stocktwits,twitter_withretweets)</param>
+        public PsychSignalDataDownloader(string rawDataDestination, string apiKey, string dataSource)
+        {
+            _rawDataDestination = rawDataDestination;
+
+            _dataSource = dataSource;
+            _apiKey = apiKey;
+        }
+        
+        /// <summary>
+        /// Download the data from the given starting date to the ending date.
+        /// Note that if the ending date is in the same hour as the current time,
+        /// we will lower the <paramref name="endDateUtc" /> by one hour in order
+        /// to make sure that we only download complete, non-changing dataa
+        /// </summary>
+        /// <param name="startDateUtc">Starting date</param>
+        /// <param name="endDateUtc">Ending date</param>
+        public void Download(DateTime startDateUtc, DateTime endDateUtc)
+        {
+            if (startDateUtc < endDateUtc.AddDays(-15))
+            {
+                throw new ArgumentException("The starting date can only be at most 15 days from now");
+            }
+
+            Directory.CreateDirectory(_rawDataDestination);
+            var now = DateTime.UtcNow;
+
+            // Makes sure we only get final, non-changing data by checking if the current hour matches the 
+            // hour of the upper date bound and subtracting an hour from it if that is the case
+            if (new DateTime(now.Year, now.Month, now.Day, now.Hour, 0, 0) == new DateTime(endDateUtc.Year, endDateUtc.Month, endDateUtc.Day, endDateUtc.Hour, 0, 0))
+            {
+                endDateUtc = endDateUtc.AddHours(-1);
+            }
+            
+            // PsychSignal paginates data by hour
+            for (; startDateUtc < endDateUtc; startDateUtc = startDateUtc.AddHours(1))
+            {
+                var rawDataPath = Path.Combine(_rawDataDestination, $"{startDateUtc:yyyyMMdd_HH}_{_dataSource}.csv");
+                var rawDataPathTemp = Path.Combine(Path.GetTempPath(), $"{startDateUtc:yyyyMMdd_HH}_{_dataSource}.csv.tmp");
+                
+                // Don't download files we already have
+                if (File.Exists(rawDataPath))
```

Comment:
Is there any concern regarding new data in these files? I recall the data is updated hourly? -- also looks like the hours are included in the file path. Is there any chance of downloading a _partial_ hour? For example, consider the current time is 10:30AM, is it possible to request 10AM->11AM  and only get the first 30 minutes? If so, this check will prevent us from downloading the full hour's worth of data when it's available.

Reply:
Yes, it is possible to download a partial hour. However, try to make sure that doesn't happen with the following lines:

https://github.com/QuantConnect/Lean/blob/42980c5094999f0f4ca7d2f3f61ec97b7af95efb/ToolBox/PsychSignalDataConverter/PsychSignalDataDownloader.cs#L79-L82

But on second thought, it's a bit naive to assume that the top bound will always be `DateTime.UtcNow` - I'll probably have to do some refactoring here to avoid that issue. Thanks for pointing this out :smile:

Reply:
note to  future readers -- this check is now safe given the protections against downloading partial data provided [here](https://github.com/QuantConnect/Lean/pull/3349/files#diff-4195c27ad597116ef70315139cfaf2caR87).